File: 853-most-profit-assigning-work/most-profit-assigning-work.java
class Solution {
    public int maxProfitAssignment(int[] difficulty, int[] profit, int[] worker) {
        List<int[]>vc = new ArrayList<>();
        int n =difficulty.length;
        for(int i=0;i<n;i++){
            vc.add(new int[]{difficulty[i],profit[i]});
        }
        Collections.sort(vc,(a,b)->a[0]-b[0]);
        Arrays.sort(worker);
        PriorityQueue<int[]>maxprof = new PriorityQueue<>((a,b)->b[0]-a[0]);
        int total = 0;
        int idx = 0;
        for(int i:worker){
            while(idx<n && i>=vc.get(idx)[0]){
                maxprof.offer(new int[]{vc.get(idx)[1],vc.get(idx)[0]});
                idx++;
            }
            if(!maxprof.isEmpty()){
                total+=maxprof.peek()[0];
            }
        }
        return total;
    }
}

File: 853-most-profit-assigning-work/most-profit-assigning-work.cpp
typedef pair<int,int> pi;
class Solution {
public:
    int maxProfitAssignment(vector<int>& difficulty, vector<int>& profit, vector<int>& worker) {
       int maxdiff = *max_element(difficulty.begin(),difficulty.end());
       vector<int>maxprof(maxdiff+1,0);
       for(int i=0;i<difficulty.size();i++){
        maxprof[difficulty[i]]=max(maxprof[difficulty[i]],profit[i]);
       }
       for(int i=1;i<=maxdiff;i++){
        maxprof[i]=max(maxprof[i],maxprof[i-1]);
       }
       int cnt = 0;
       for(int i:worker){
        if(i>=maxdiff){
            cnt+=maxprof[maxdiff];
        }
        else cnt+=maxprof[i];
       }
       return cnt;
    }
};